File: OnlinerMoto/OnlinerMotoVehicleDetailsViewController.m
//
//  OnlinerMotoVehicleDetailsViewController.m
//  OnlinerMoto
//
//  Created by Igor Karpov on 17.11.2013.
//  Copyright (c) 2013 KarpovIV. All rights reserved.
//

#import "OnlinerMotoVehicleDetailsViewController.h"
#import "VehicleItem.h"
#import "VehicleItemDetails.h"
#import "VehicleItemsRepositoryProtocol.h"
#import "VehicleItemsProviderProtocol.h"
#import "OnlinerMotoPhotoCell.h"
#import "LoadingIndicatorView.h"

@interface OnlinerMotoVehicleDetailsViewController()
{
    VehicleItemDetails *_vehicleItemDetails;
}
@end

@implementation OnlinerMotoVehicleDetailsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.nameLabel.text = self.vehicleItem.name;
    self.priceLabel.text = [NSString stringWithFormat:@"$%u", self.vehicleItem.price];
    self.yearLabel.text =[NSString stringWithFormat:@"%u", self.vehicleItem.year];
    self.mileageLabel.text =[NSString stringWithFormat:@"%u km", self.vehicleItem.mileage];
    self.briefDescriptionLabel.text = self.vehicleItem.briefDescription;
    self.urlLabel.text = self.vehicleItem.detailsUrl;
    
    [self.photosCollectionView setBackgroundColor:[UIColor whiteColor]];
    
    [self.addVehicleItemToTaggedButton setEnabled:self.isTaggingAvailable];
    
    [self showLoadingIndicators];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        
        _vehicleItemDetails = [self.vehicleItemsProvider getItemDetailsForItem:self.vehicleItem];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self hideLoadingIndicators];
            
            [self showVehicleItemDetails:_vehicleItemDetails];
        });
    });
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)addVehicleItemToTagged:(id)sender
{
    [self.vehicleItemsRepository addVehicleItem:self.vehicleItem];
    [self.addVehicleItemToTaggedButton setEnabled:NO];
    
    [OnlinerMotoVehicleDetailsViewController showMessage:@"Item marked as tagged"];
}

+ (void)showMessage:(NSString *)message
{
    [[[UIAlertView alloc] initWithTitle:nil
                                message:message
                               delegate:nil
                      cancelButtonTitle:@"OK"
                      otherButtonTitles:nil] show];
}

- (IBAction)hideDetailsView:(id)sender
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - UICollectionView Datasource

- (NSInteger)collectionView:(UICollectionView *)view numberOfItemsInSection:(NSInteger)section
{
   if (_vehicleItemDetails == nil
        || _vehicleItemDetails.allPhotos == nil)
    {
        return 0;
    }
    
    return [_vehicleItemDetails.allPhotos count];
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)cv cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    OnlinerMotoPhotoCell *cell = (OnlinerMotoPhotoCell *)[cv dequeueReusableCellWithReuseIdentifier:@"OnlinerMotoPhotoCell" forIndexPath:indexPath];
    
    cell.photoImage.image = [UIImage imageWithData:_vehicleItemDetails.allPhotos[indexPath.row]];
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath;
{
    if (indexPath.section == 0 && indexPath.row == 1)
    {
        NSString *text = _vehicleItemDetails.vehicleItem.briefDescription;
        CGSize constraint = CGSizeMake(320.0f - (20.0f * 2), 20000.0f);
        CGSize size = [text sizeWithFont:[UIFont systemFontOfSize:15.0f] constrainedToSize:constraint lineBreakMode:NSLineBreakByWordWrapping];
        CGFloat height = MAX(size.height, 44.0f);
        
        return height + 10.0f;
    }
    else if (indexPath.section == 1)
    {
        return 320.0f;
    }
    else if (indexPath.section == 2)
    {
        NSString *text = _vehicleItemDetails.additionalDescription;
        CGSize constraint = CGSizeMake(320.0f - (20.0f * 2), 20000.0f);
        CGSize size = [text sizeWithFont:[UIFont systemFontOfSize:17.0f] constrainedToSize:constraint lineBreakMode:NSLineBreakByWordWrapping];
        CGFloat height = MAX(size.height, 44.0f);
        
        return height + 10.0f;
    }
    else
    {
        return 44.0f;
    }
}

#pragma mark -- UIAlertView

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark -- Private Mehtods

- (void)showLoadingIndicators
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
    [self.tableView setScrollEnabled:NO];
    [self.view addSubview:[[LoadingIndicatorView alloc] initWithFrame:self.view.bounds]];
}

- (void)hideLoadingIndicators
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    
    for (UIView *subView in self.view.subviews)
    {
        if ([subView isKindOfClass:[LoadingIndicatorView class]])
        {
            [subView removeFromSuperview];
        }
    }
    [self.tableView setScrollEnabled:YES];
}

- (void)showVehicleItemDetails:(VehicleItemDetails *)itemDetails
{
    if (!itemDetails)
    {
        [[[UIAlertView alloc] initWithTitle:nil
                                    message:@"Sorry, this item doesn't exist anymore."
                                   delegate:self
                          cancelButtonTitle:@"OK"
                          otherButtonTitles:nil] show];        
        return;
    }
    
    self.locationLabel.text = itemDetails.location;
    self.additionalDescriptionLabel.text = itemDetails.additionalDescription;
    self.additionalDescriptionLabel.numberOfLines = 0;
    [self.additionalDescriptionLabel sizeToFit];
    [self.briefDescriptionLabel sizeToFit];
       
    [self.photosCollectionView reloadData];
    [self.tableView reloadData];
}

@end

File: OnlinerMoto/OnlinerMotoVehicleDetailsViewController.swift
//
//  OnlinerMotoVehicleDetailsViewController.swift
//  OnlinerMoto
//
//  Created by Igor Karpov on 15.11.2014.
//  Copyright (c) 2014 KarpovIV. All rights reserved.
//

import Foundation

class OnlinerMotoVehicleDetailsViewController : UITableViewController, UICollectionViewDataSource
{
    @IBOutlet weak private var nameLabel: UILabel!
    @IBOutlet weak private var priceLabel: UILabel!
    @IBOutlet weak private var yearLabel: UILabel!
    @IBOutlet weak private var mileageLabel: UILabel!
    @IBOutlet weak private var briefDescriptionLabel: UILabel!
    @IBOutlet weak private var locationLabel: UILabel!
    @IBOutlet weak private var urlLabel: UILabel!
    @IBOutlet weak private var additionalDescriptionLabel: UILabel!
    
    @IBOutlet weak private var photosCollectionView: UICollectionView!
    @IBOutlet weak private var addVehicleItemToTaggedButton: UIBarButtonItem!
    
    var vehicleItem: VehicleItem!
    var isTaggingAvailable: Bool = false
    private var vehicleItemDetails: VehicleItemDetails?
    
    required init(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
    }

    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        self.nameLabel.text = self.vehicleItem.name
        self.priceLabel.text = "$\(self.vehicleItem.price)"
        self.yearLabel.text = String(self.vehicleItem.year)
        self.mileageLabel.text = "\(self.vehicleItem.mileage) km"
        self.briefDescriptionLabel.text = self.vehicleItem.briefDescription
        self.urlLabel.text = self.vehicleItem.detailsUrl
        
        self.photosCollectionView.backgroundColor = UIColor.whiteColor()
        self.addVehicleItemToTaggedButton.enabled = self.isTaggingAvailable
        
        showLoadingIndicators()

        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0))
        {
            var itemDetails = self.getGlobalVehicleItemsProvider().getItemDetailsForItem(self.vehicleItem)
            
            dispatch_async(dispatch_get_main_queue())
            {
                self.hideLoadingIndicators()
                self.vehicleItemDetails = itemDetails
                if (self.vehicleItemDetails != nil)
                {
                    self.showVehicleItemDetails(self.vehicleItemDetails!)
                }
            }
        }
    }
    
    @IBAction private func addVehicleItemToTagged(sender: AnyObject)
    {
        self.getGlobalVehicleItemsRepository().addVehicleItem(self.vehicleItem)
        self.addVehicleItemToTaggedButton.enabled = false
        
        OnlinerMotoVehicleDetailsViewController.showMessage("Item marked as tagged")
    }
    
    private class func showMessage(message: String)
    {
        UIAlertView(
                title: "",
                message: message,
                delegate: nil,
                cancelButtonTitle: "OK").show()
    }
    
    @IBAction private func hideDetailsView(sender: AnyObject)
    {
        dismissViewControllerAnimated(true, completion: nil)
    }
    
    func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int
    {
        var number = 0
        
        if let photos = self.vehicleItemDetails?.allPhotos
        {
            number = photos.count
        }
        
        return number
    }
    
    func collectionView(collectionView: UICollectionView,
        cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell
    {
        var cell = collectionView.dequeueReusableCellWithReuseIdentifier(
            "OnlinerMotoPhotoCell",
            forIndexPath: indexPath) as OnlinerMotoPhotoCell
        
        if let photos = self.vehicleItemDetails?.allPhotos
        {
            cell.photoImage.image = UIImage(data: photos[indexPath.row])
        }
        
        return cell
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat
    {
        var height: CGFloat = 44
        if (indexPath.section == 0 && indexPath.row == 1)
        {
            if let text: String = self.vehicleItemDetails?.vehicleItem?.briefDescription
            {
                var foundationText = text as NSString
                var constraint = CGSizeMake(320 - (20 * 2), 20000)
                var pstyle = NSMutableParagraphStyle()
                pstyle.lineBreakMode = .ByWordWrapping
                
                var attributes = [NSFontAttributeName: UIFont.systemFontOfSize(17.0), NSParagraphStyleAttributeName: pstyle]
                
                var rect = foundationText.boundingRectWithSize(
                    constraint,
                    options: NSStringDrawingOptions.UsesLineFragmentOrigin,
                    attributes: attributes,
                    context: nil)
                
                height = 10 + max(rect.height, 44.0)
            }
            else
            {
                height = 10
            }
        }
        else if (indexPath.section == 1)
        {
            height = 320
        }
        else if (indexPath.section == 2)
        {
            if let text: String = self.vehicleItemDetails?.additionalDescription
            {
                var foundationText = text as NSString
                var constraint = CGSizeMake(320 - (20 * 2), 20000)
                var pstyle = NSMutableParagraphStyle()
                pstyle.lineBreakMode = .ByWordWrapping
                
                var attributes = [NSFontAttributeName: UIFont.systemFontOfSize(17.0), NSParagraphStyleAttributeName: pstyle]
                
                var rect = foundationText.boundingRectWithSize(
                    constraint,
                    options: NSStringDrawingOptions.UsesLineFragmentOrigin,
                    attributes: attributes,
                    context: nil)
                
                height = 10 + max(rect.height, 44.0)
            }
            else
            {
                height = 10
            }
        }
        
        return height
    }
    
    private func showVehicleItemDetails(itemDetails: VehicleItemDetails)
    {
        self.locationLabel.text = itemDetails.location
        self.additionalDescriptionLabel.text = itemDetails.additionalDescription
        self.additionalDescriptionLabel.numberOfLines = 0
        self.additionalDescriptionLabel.sizeToFit()
        self.briefDescriptionLabel.sizeToFit()

        self.photosCollectionView.reloadData()
        self.tableView.reloadData()
    }
    
    private func showLoadingIndicators()
    {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = true
        self.tableView.scrollEnabled = false
        self.view.addSubview(LoadingIndicatorView(frame: self.view.bounds))
    }
    
    private func hideLoadingIndicators()
    {
        UIApplication.sharedApplication().networkActivityIndicatorVisible = false
        
        for subview in self.view.subviews
        {
            if (subview is LoadingIndicatorView)
            {
                subview.removeFromSuperview()
            }
        }
        self.tableView.scrollEnabled = true
    }
    
    private func getGlobalVehicleItemsRepository() -> VehicleItemsRepositoryProtocol
    {
        var appDelegate = UIApplication.sharedApplication().delegate as OnlinerMotoAppDelegateProtocol
        return appDelegate.vehicleItemsRepository
    }
    
    private func getGlobalVehicleItemsProvider() -> VehicleItemsProviderProtocol
    {
        var appDelegate = UIApplication.sharedApplication().delegate as OnlinerMotoAppDelegateProtocol
        return appDelegate.vehicleItemsProvider
    }
}